Non-forking endpoint adapter [ESD-103]

silverjam opened the pull request:
To improve the efficiency of the endpoint adapter (and to clean up the process table) we change to using an event loop from `pk_loop_t` in order to read and service data in `endpoint_adapter`.  This removes 3 processes per adapter in the system, but leaves a "fork per client" for the TCP server.  We can eventually remove the fork from the TCP server as well.

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2018120522-9-g2e106882

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=802063)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018120522-9-g2e106882&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018120522-9-g2e106882&build_type=buildroot_pr&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2018120522-9-g2e106882
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018120522-9-g2e106882)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018120522-9-g2e106882/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018120522-9-g2e106882

silverjam commented:
@yizow Can you test CAN support on this branch?

pmiettinen commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -128,11 +138,25 @@ static const char *udp_connect_addr = NULL;
+} read_ctx = {
+  .buffer = NULL,
+  .fill = 0,
 };
```
`.size = 0,`?

silverjam replied:
fixed

pmiettinen commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -56,15 +56,27 @@ static pk_metrics_t *MR = NULL;
+  PK_METRICS_ENTRY("tx/write/count",           "per_second",     M_U32,         M_UPDATE_COUNT,   M_RESET_DEF, tx_write_count),
+  PK_METRICS_ENTRY("tx/write/size/per_second", "total",          M_U32,         M_UPDATE_SUM,     M_RESET_DEF, tx_write_size_total),
+  PK_METRICS_ENTRY("tx/write/size/per_second", "average",        M_U32,         M_UPDATE_AVERAGE, M_RESET_DEF, tx_write_size_average,
+                   M_AVERAGE_OF(MI,         tx_write_size_total, tx_write_count))
```
Can anything be done to this formatting? At least github is having trouble showing it.

pmiettinen commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -586,30 +615,69 @@ static ssize_t fd_write(int fd, const void *buffer, size_t count)
+  memcpy(read_ctx.buffer, buff, length);
+  read_ctx.fill += length;
+
+  // Return -1 to terminate the read loop, only read one packet
```
comment style

silverjam replied:
Fixed

pmiettinen commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -698,287 +766,256 @@ static ssize_t handle_write_all_via_framer(handle_t *handle,
-          debug_printf("handle_init for sub returned error\n");
-          exit(EXIT_FAILURE);
-        }
+  void *handle = pk_loop_timer_add(loop_ctx.loop, 1000, do_metrics_flush, NULL);
```
Is this unused variable or don't I just see it..

silverjam replied:
Only purpose is for the assert on the next line

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2018121420-23-gd7650a95

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=819822)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018121420-23-gd7650a95&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018121420-23-gd7650a95&build_type=buildroot_pr&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2018121420-23-gd7650a95
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018121420-23-gd7650a95)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018121420-23-gd7650a95/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018121420-23-gd7650a95

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2018121420-27-g90b1273d

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=820665)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018121420-27-g90b1273d&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018121420-27-g90b1273d&build_type=buildroot_pr&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2018121420-27-g90b1273d
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018121420-27-g90b1273d)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018121420-27-g90b1273d/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018121420-27-g90b1273d

silverjam commented:
TODO:
- [ ] @yizow: retest that CAN works
- [ ] Get more review comments (@yiiiyang2010 @pmiettinen @yizow @benjaminaltieri)?

denniszollo commented:
Will this allow us to divorce in and out communication modes on the interface eventually ? (i.e. RTCMv3 In, NEMA out) ?
I wouldn't want to refactor things in such a way that this common customer request use case gets harder.

silverjam commented:
@denniszollo I don't think this change will make that any more/less hard

pmiettinen commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -1,6 +1,6 @@
 /*
- * Copyright (C) 2016 Swift Navigation Inc.
- * Contact: Jacob McNamee <jacob@swiftnav.com>
+ * Copyright (C) 2014-2018 Swift Navigation Inc.
```
2014? Seems cheating.

pmiettinen replied:
And update to 2019 where applicable?

silverjam replied:
fixed

pmiettinen commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -56,15 +56,27 @@ static pk_metrics_t *MR = NULL;
 // clang-format off
```
c89

silverjam replied:
fixed

pmiettinen commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -698,287 +767,256 @@ static ssize_t handle_write_all_via_framer(handle_t *handle,
+static void io_loop_pubsub(pk_loop_t *loop, handle_t *read_handle, handle_t *write_handle)
 {
-  debug_printf("io loop begin\n");
+  if (read_handle->pk_ept != NULL) {
```
Inverting to
```C
  if (read_handle->pk_ept == NULL) {
    PK_METRICS_UPDATE(MR, MI.rx_read_count);
  } else {
    PK_METRICS_UPDATE(MR, MI.tx_read_count);
  }
```
would save one logical hop for us poor souls with limited brain power.

silverjam replied:
fixed see UPDATE_IO_LOOP_METRIC macro

pmiettinen commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -698,287 +767,256 @@ static ssize_t handle_write_all_via_framer(handle_t *handle,
+  }
 
-    PK_METRICS_UPDATE(MR, MI.read_count);
+  if (read_handle->pk_ept != NULL) {
```
Same here.

silverjam replied:
fixed, added a helper macro since I forgot what the different between the 2 cases was

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2019011120-21-gd754dbd2

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=854605)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-21-gd754dbd2&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-21-gd754dbd2&build_type=buildroot_pr&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2019011120-21-gd754dbd2
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019011120-21-gd754dbd2)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019011120-21-gd754dbd2/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019011120-21-gd754dbd2

silverjam commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -351,6 +376,11 @@ static int parse_options(int argc, char *argv[])
     }
   }
 
+  if (port_name == NULL) {
```
fixed

silverjam commented on package/endpoint_adapter/src/endpoint_adapter.c:
```diff
@@ -698,287 +767,256 @@ static ssize_t handle_write_all_via_framer(handle_t *handle,
+  piksi_log(LOG_ERR | LOG_SBP, error);
+  fprintf(stderr, "%s\n", error);
+
+  exit(EXIT_FAILURE);
```
fixed

silverjam commented:
@yizow @pmiettinen @benjaminaltieri This is probably worth another once over, at least in `endpoint_adapter_can.c`, I fixed the issue with CAN, I haven't been able to root cause the issue (to determine if it's an issue with SocketCAN or with libuv)... However, as a result of the fixes I moved a lot of the CAN specific stuff out of endpoint_adapter and in to the `endpoint_adapter_can` module.  Using threads and pipes to work around the issue is a little gross, but it works.

pmiettinen commented:
Well, if it works.. Seems ok.

pmiettinen commented on package/endpoint_adapter/src/endpoint_adapter_can.c:
```diff
@@ -10,17 +10,189 @@
  * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
  */
 
+#define _GNU_SOURCE
```
What requires this? I have seen comments that this is a hit for portability.

silverjam replied:
The `pipe2` call with O_DIRECT which enables packet style writes/reads from the pipe.

silverjam commented:
Bench-marking this doesn't show a significant difference in SBP fileio performance.

silverjam commented:
Potential commit message:

```
## Design notes

* Make endpoint_adapter non-block:

To improve the efficiency of the endpoint adapter (and to clean up the process table) we change to using an event loop from pk_loop_t in order to read and service data in endpoint_adapter. This removes 3 processes per adapter in the system, but leaves a "fork per client" for the TCP server. We can eventually remove the fork from the TCP server as well.

* Use a pair of helper threads for CAN

Since CAN doesn't seem to work in O_NONBLOCK mode with libuv (we haven't
found the root cause of this yet).  Work around the issue by using a
pair of helper threads to read/write data from/to the SocketCAN handle.

The read thread does a "non-blocking" select to poll the descriptor to
see if it's ready, then sleeps for 1ms if it's not.  CAN data that's
read is written to the "write" end of the CAN read data pipe.

The write thread does a normal select on the read end of the write pipe
waiting for data from the main endpoint_adapter thread.  Data that
arrives is sent over CAN via a blocking write.

## Testing
Informal benchmarking using `watch -n1 'find /var/log/metrics/sbp_fileio/external -type f -print -exec cat {} \;'` showed maybe a slight slowdown with the non-blocking `endpoint_adapter` but the different was only 2-5%
```

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2019011120-40-g2e4fa729

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=860134)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-40-g2e4fa729&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-40-g2e4fa729&build_type=buildroot_pr&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2019011120-40-g2e4fa729
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019011120-40-g2e4fa729)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019011120-40-g2e4fa729/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019011120-40-g2e4fa729

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2019011120-45-gea604b0f

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=861549)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-45-gea604b0f&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-45-gea604b0f&build_type=buildroot_pr&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2019011120-45-gea604b0f
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019011120-45-gea604b0f)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019011120-45-gea604b0f/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019011120-45-gea604b0f

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2019011120-46-gfa5361eb

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=861649)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-46-gfa5361eb&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-46-gfa5361eb&build_type=buildroot_pr&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2019011120-46-gfa5361eb
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019011120-46-gfa5361eb)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019011120-46-gfa5361eb/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019011120-46-gfa5361eb

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2019011120-47-g911d72d4

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=862860)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-47-g911d72d4&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2019011120-47-g911d72d4&build_type=buildroot_pr&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2019011120-47-g911d72d4
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2019011120-47-g911d72d4)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2019011120-47-g911d72d4/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2019011120-47-g911d72d4
